### ISSUE TITLE
many issues since upgrading to 1.16 around @nexrender/action-encode

### ISSUE THREAD

**Issue by FinnFrotscher** [inlife/nexrender#389]
i upgraded to 1.16, and since then i am having a host of problems around encode.
i am running the latest macos.

it starts with requiring the module. this error is thrown only after the render is done which is particulary annoying since i have to render an entire video (~5 minutes) to test this issue.
```
not require module '@nexrender/action-encode'"
{
  "nested": {
    "code": "MODULE_NOT_FOUND",
    "requireStack": [
      "/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/node-fetch-progress/dist/index.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/action-encode/index.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/requireg/lib/requireg.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/core/src/tasks/download.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/core/src/index.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/src/templates/interview/steps/render.js",
      "/Users/finnfrotscher/code/cofenster/api.renderer/src/index.js"
    ]
  }
}
lifecycle error NestedError: Could not require module '@nexrender/action-encode'
    at requireg (/Users/finnfrotscher/code/cofenster/api.renderer/node_modules/requireg/lib/requireg.js:14:11)
    at /Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/core/src/tasks/actions.js:24:16
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
Caused By: Error: Cannot find module 'date-fns/distance_in_words_to_now'
Require stack:
- /Users/finnfrotscher/code/cofenster/api.renderer/node_modules/node-fetch-progress/dist/index.js
- /Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/action-encode/index.js
- /Users/finnfrotscher/code/cofenster/api.renderer/node_modules/requireg/lib/requireg.js
- /Users/finnfrotscher/code/cofenster/api.renderer/node_modules/@nexrender/core/src/tasks/download.js
...
```
... i have it installed...
this is my `package.json`
```
    "@nexrender/action-copy": "^1.8.16",
    "@nexrender/action-encode": "^1.16.0",
    "@nexrender/action-upload": "^1.10.0",
    "@nexrender/core": "^1.16.0",
    "@nexrender/provider-s3": "^1.8.16",
```
----

If i install action-encode manually,
I am running into issues with ffmpeg, again, after having to render an entire video:

When running as `sudo`
```
MCsc0M7y0Gx0VAi97CNuu] action-encode: output file /Users/finnfrotscher/code/cofenster/api.renderer/tmp/MCsc0M7y0Gx0VAi97CNuu/encoded.mp4
> ffmpeg binary b4.2.2 is not found
> downloading a new ffmpeg binary b4.2.2 to: /Users/finnfrotscher/code/cofenster/api.renderer/tmp/ffmpeg-b4.2.2
> ffmpeg binary b4.2.2 was successfully downloadednds

[ERROR]

"Error loading postrender module @nexrender/action-encode: Error: 
Error starting ffmpeg process: 
Error: spawn /Users/finnfrotscher/code/cofenster/api.renderer/tmp/ffmpeg-b4.2.2 EACCES"
{}

```
when running without `sudo`:
```
"Running Script: /Users/finnfrotscher/code/cofenster/api.renderer/tmp/kqN3wKxjm_IjZdjNRmP5L/nexrender-kqN3wKxjm_IjZdjNRmP5L-script.jsx
PROGRESS: Adding specified comp to Render Queue
aerender Error: After Effects error: 
Error in output for render queue item 1, output module 1. 
Can not create a file in directory /Users/finnfrotscher/code/cofenster/api.renderer/tmp/kqN3wKxjm_IjZdjNRmP5L. 

Try checking write permissions.
aerender version 17.0.6x35
PROGRESS: Launching After Effects...
PROGRESS: ...After Effects successfully launched
{
```

first off: ` /Users/finnfrotscher/code/cofenster/api.renderer/tmp/ffmpeg-b4.2.2` is not the desired location.
second: i cant run this with and without sudo. what do i do?

**Comment by FinnFrotscher** [comment]
temporary fix is to run `sudo chmod +x ffmpeg-xxx`
ffmpeg is installed globally using brew. its the newest version.

**Comment by FinnFrotscher** [comment]
now i have to reinstall ffmpeg every time i trash my tmp folder. and render the first project twice.

**Comment by inlife** [comment]
Could you try out `v1.16.1`?
I added an automatic `chmod` setting, that should resolve issues with permissions.

**Comment by FinnFrotscher** [comment]
will do soon.

**Comment by FinnFrotscher** [comment]
still having an issue.
`@nexrender/action-encode` is not installed when running `yarn`. i have to install it manually after every `rm -rf node_modules`

**Comment by inlife** [comment]
Oh, hold on, I thought you were using binaries.
Could you please share your setup info, how packages are installed, and how are they used?
Possibly with more details, I could figure out what is going on.

**Comment by FinnFrotscher** [comment]
allright, i will try to. if i am missing something, let me know:
some of my dependencies:
```
  "dependencies": {
    "@babel/cli": "^7.8.4",
    "@babel/core": "^7.8.4",
    "@babel/node": "^7.8.4",
    "@babel/plugin-proposal-class-properties": "^7.8.3",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/plugin-proposal-object-rest-spread": "^7.8.3",
    "@babel/plugin-transform-runtime": "^7.8.3",
    "@babel/polyfill": "^7.8.3",
    "@babel/preset-env": "^7.8.4",
    "@babel/register": "^7.8.3",
    "@babel/runtime": "^7.8.4",
    "@nexrender/action-copy": "^1.8.16",
    "@nexrender/action-encode": "^1.16.1",
    "@nexrender/action-upload": "^1.10.0",
    "@nexrender/core": "^1.16.0",
    "@nexrender/provider-s3": "^1.8.16",
    "apollo-boost": "^0.4.7",
    "babel-jest": "^25.1.0",
    "babel-plugin-transform-runtime": "^6.23.0",
    "dotenv": "^7.0.0",
    "graphql": "^14.6.0",
    "lodash": "^4.17.15",
    "nodemon": "^1.19.4",
  },
```
my service is build on node, run by nodemon and compiled with babel.
Its basically a large version of the test project.
the service receives a data object, and parses it into instructions for nexrender.
nexrender runs the aerender binary.
The aerender is installed by the creative cloud installer.

**Comment by inlife** [comment]
are you installing the @nexrender/action-encode module globally?

**Comment by FinnFrotscher** [comment]
the `ffmpeg` issue is solved.
`nexrender/action-encode` still needs to be installed manually after each module wipe.
i am not installing action encode globally.

**Comment by inlife** [comment]
I actually noticed there is a bit weird reason for that error in the main message:
![image](https://user-images.githubusercontent.com/2182108/83231389-16532180-a194-11ea-98ca-6a78898e3676.png)

I have a theory that somehow that package is not included properly, added it as a direct dependency. Published as `v1.16.3`, please try it out.

**Comment by Arcetrix** [comment]
Hi,

i installed the binary last version 1.16.3 of worker .exe but it seems don't solve the error below:
[vOPPZXymiPfSlIGqMTLjb] starting action-encode action (ffmpeg)
[vOPPZXymiPfSlIGqMTLjb] action-encode: input file C:\Users\IOL\AppData\Local\Temp\nexrender\vOPPZXymiPfSlIGqMTLjb\result.avi
[vOPPZXymiPfSlIGqMTLjb] action-encode: output file C:\Users\IOL\AppData\Local\Temp\nexrender\vOPPZXymiPfSlIGqMTLjb\encoded.mp4
> using an existing ffmpeg binary b4.2.2 at: C:\Users\IOL\AppData\Local\Temp\nexrender\ffmpeg-b4.2.2.exe
[vOPPZXymiPfSlIGqMTLjb] error occurred: Error: Error loading postrender module @nexrender/action-encode: Error: Error in action-encode module (ffmpeg) Error: spawn UNKNOWN
    at C:\snapshot\nexrender\packages\nexrender-worker\node_modules\@nexrender\core\src\tasks\actions.js
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async start (C:\snapshot\nexrender\packages\nexrender-worker\src\index.js)

Please have you some suggestion on how to solve?

thank you

**Comment by inlife** [comment]
@Arcetrix does it work for you with an older version, `1.16.2` let's say?

**Comment by Arcetrix** [comment]
@inlife Hi, we solved installing a new version of ffmpeg (ffmpeg version git-2020-05-28-c0f01ea Copyright (c) 2000-2020 the FFmpeg developers) and copying and renaiming in '....\AppData\Local\Temp\nexrender\ffmpeg-b4.2.2.exe'. Now the .mp4 is generated and copied to the destination directory. Please do you know how can specify the pathname of the ffmpeg? or is this hardcoded to ffmpeg-b4.2.2? thank you.

**Comment by FinnFrotscher** [comment]
still riddled with errors:
```
[ObNygA1WCWJANcbU5Y36b] action-encode: output file /Users/finnfrotscher/code/cofenster/api.renderer/build/tmp/ObNygA1WCWJANcbU5Y36b/encoded.mp4
> using an existing ffmpeg binary b4.2.2 at: /Users/finnfrotscher/code/cofenster/api.renderer/build/tmp/ffmpeg-b4.2.2





[ERROR]

"Error loading postrender module @nexrender/action-encode: Error: Error in action-encode module (ffmpeg) code : null"
{}





[ERROR]

"Error loading postrender module @nexrender/action-encode: Error: Error in action-encode module (ffmpeg) code : null"
{}
```

**Comment by FinnFrotscher** [comment]
i reset my node modules and deleted the tmp folder. now its working.

**Comment by inlife** [comment]
@Arcetrix no, currently you can't provide a custom binary to the encode module, which is actually a shame, it seems to me. Probably something that will need to be added to the module in the future.

@FinnFrotscher How is it going, does it still work, are there any issues?

**Comment by FinnFrotscher** [comment]
yes. still running into this  at a few points. the latest is
```
npx @nexrender/worker --host https://nexrender.com                                                                                                                                              

npx: installed 28 in 1.662s
> starting nexrender-worker endpoint https://nexrender.com; using secret: no
using automatically determined directory of After Effects installation:
 - /Applications/Adobe After Effects 2020/aerender
adding default render-only-node licenses for After Effects at:
 - /Users/finnfrotscher/Documents/ae_render_only_node.txt
 - /Users/finnfrotscher/Documents/Adobe/ae_render_only_node.txt
checking After Effects command line renderer patch...
command line patch already is in place
[mu7DsTeuu1U-5fmWFrFvP] setting up job...
[mu7DsTeuu1U-5fmWFrFvP] working directory is: /var/folders/ym/08mhwl8174df_zdlw2m_j6280000gn/T/nexrender/mu7DsTeuu1U-5fmWFrFvP
[mu7DsTeuu1U-5fmWFrFvP] downloading assets...
[mu7DsTeuu1U-5fmWFrFvP] applying prerender actions...
[mu7DsTeuu1U-5fmWFrFvP] running script assemble...
[mu7DsTeuu1U-5fmWFrFvP] rendering job...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 0%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 1%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 15%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 29%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 43%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 58%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 72%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 86%...
[mu7DsTeuu1U-5fmWFrFvP] rendering progress 100%...
[mu7DsTeuu1U-5fmWFrFvP] rendering took ~23.046 sec.
[mu7DsTeuu1U-5fmWFrFvP] writing aerender job log to: /var/folders/ym/08mhwl8174df_zdlw2m_j6280000gn/T/nexrender/aerender-mu7DsTeuu1U-5fmWFrFvP.log
[mu7DsTeuu1U-5fmWFrFvP] applying postrender actions...
[mu7DsTeuu1U-5fmWFrFvP] error occurred: NestedError: Could not require module '@nexrender/action-encode'
    at requireg (/Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/node_modules/requireg/lib/requireg.js:14:11)
    at /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/node_modules/@nexrender/core/src/tasks/actions.js:24:16
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
Caused By: Error: Cannot find module 'date-fns/distance_in_words_to_now'
Require stack:
- /Users/finnfrotscher/.config/yarn/global/node_modules/node-fetch-progress/dist/index.js
- /Users/finnfrotscher/.config/yarn/global/node_modules/@nexrender/action-encode/index.js
- /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/node_modules/requireg/lib/requireg.js
- /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/node_modules/@nexrender/core/src/tasks/download.js
- /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/node_modules/@nexrender/core/src/index.js
- /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/src/index.js
- /Users/finnfrotscher/.npm/_npx/4357/lib/node_modules/@nexrender/worker/src/bin.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:957:15)
    at Function.Module._load (internal/modules/cjs/loader.js:840:27)
    at Module.require (internal/modules/cjs/loader.js:1019:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/Users/finnfrotscher/.config/yarn/global/node_modules/node-fetch-progress/dist/index.js:3:56)
    at Module._compile (internal/modules/cjs/loader.js:1133:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1153:10)
    at Module.load (internal/modules/cjs/loader.js:977:32)
    at Function.Module._load (internal/modules/cjs/loader.js:877:14)
    at Module.require (internal/modules/cjs/loader.js:1019:19)
```

**Comment by Arcetrix** [comment]
Hi @inlife ,

S.O. Windows 10.
My 'worker' run on C./ drive and i need to copy the video file .mp4 to a different drive let say Z:\ (a samba shared drive).
I have tried in many ways to do that setting the path value for 'output' to:
"output": "Z:/bigBunny7.mp4"
or
"output": "file:///Z:/bigBunny7.mp4"
or (relative path):
"output": "../../../Z:/bigBunny7.mp4"

but the worker return error.

Have you some suggestions about this problem?

Below the portion of JSON:
    "actions":{
        "postrender": [
            {
                "module": "@nexrender/action-encode",
                               "output": "encoded.mp4",
                "preset": "mp4",
                               "params": {"-vcodec": "libx264", "-r": 25}
            },
            {
                "module": "@nexrender/action-copy",
                "input":  "encoded.mp4",
                "output": "Z:/bigBunny7.mp4"

Thank you

**Comment by FinnFrotscher** [comment]
@Arcetrix thats unrelated to this issue. open a seperate issue, please

**Comment by Arcetrix** [comment]
Hi @FinnFrotscher , sorry i opened a separate issue (https://github.com/inlife/nexrender/issues/404).

**Comment by inlife** [comment]
@FinnFrotscher it does not make sense to me why would it sometimes error out with unexisting module
like something is removing it

**Comment by FinnFrotscher** [comment]
i dont know. 
i am running the command on npx, it should be the latest version of the worker.
it may be that the import fails because of an error with date-fns
```
Caused By: Error: Cannot find module 'date-fns/distance_in_words_to_now'
```

**Comment by inlife** [comment]
Yes, that is what I'm talking about. This module (`date-fns`) is installed alongside the nexrender-action-encode, but somehow it's like it disappears after a few working attempts, that is what I understood from your words.

**Comment by FinnFrotscher** [comment]
its working when i use the binary.
is it possible that `npx @nexrender/worker` fetches a  version of action-encode that is out of date?

**Comment by FinnFrotscher** [comment]
https://github.com/inlife/nexrender/blob/master/packages/nexrender-worker/package.json#L20
i guess its an outdated package reference.
If you bump this to 1.16.3 it should be fixed

**Comment by inlife** [comment]
Oke, that is something we can easily test out.

**Comment by inlife** [comment]
And unfortunately that didn't work, Travis reported an error that package with such version does not exist.

**Comment by FinnFrotscher** [comment]
Can you or anyone  reproduce  the issue?

**Comment by FinnFrotscher** [comment]
Still facing this issue in production.
package.json
```
 "@nexrender/action-copy": "^1.8.16",
    "@nexrender/action-encode": "^1.17.0",
    "@nexrender/action-upload": "^1.10.0",
    "@nexrender/api": "^1.14.4",
    "@nexrender/core": "^1.17.1",
    "@nexrender/provider-ftp": "^1.10.0",
    "@nexrender/provider-gs": "^1.9.2",
    "@nexrender/provider-s3": "^1.8.16",
    "@nexrender/server": "^1.16.4",
```

webpack config
```
 {
    entry: path.resolve(__dirname, './src/index.js'),
    target: 'node',
    mode: isDevBuild ? 'development' : 'production',
    node: { __dirname: false },
    output: {
      path: path.resolve(__dirname, './dist'),
      filename: 'index.js',
    },
    resolve: {
      extensions: ['.js', '.mjs'],
      modules: ['node_modules', 'src'],
    },
    module: {
      rules: [
        {
          test: /\.(js?|mjs)$/,
          use: ['babel-loader'],
        },
      ],
    },
    devtool: isDevBuild && 'cheap-module-source-map',
    plugins: [
      new CleanWebpackPlugin(),
    ].filter(Boolean),
  };
```


process stdout
```
lifecycle error NestedError: Could not require module '@nexrender/action-encode'
    at requireg (/Users/finnfrotscher/code/monorepo/packages/render-builder/dist/index.js:222291:11)
    at /Users/finnfrotscher/code/monorepo/packages/render-builder/dist/index.js:81259:12
```

**Comment by FinnFrotscher** [comment]
my guess would be that the `requireq` package employed inside core/tasks/actions is not loading the module I pass. 
the package might be out of date.

**Comment by inlife** [comment]
The thing is that the module never actually required nor added to package.json as a dependency, since it is a rather plugin, that users choose to install manually. Hence the reason there might be some issues caused by that plugin-based approach in your specific case since I never actually tested nexrender with webpack. 

You can try to define `NEXRENDER_REQUIRE_PLUGINS` env var, it is something used for building process with `pkg`, it forces all manual requires to be present in the code while it does scanning to resolve them at the build time, for bundling everything together automatically
![image](https://user-images.githubusercontent.com/2182108/86877002-9dc67580-c0ee-11ea-9b24-a10d47b00761.png)

**Comment by morneluckybeard** [comment]
I'm getting a similar issue, I've got a VueJs app built with Electron and once rendering finishes and any of my actions need to run I get the same problem: 
![image](https://user-images.githubusercontent.com/58560781/95990484-df706600-0e2b-11eb-8517-be7ce7e478f7.png)

I have tried to set the environment variable, still it throws the same error. 

code: "MODULE_NOT_FOUND"
Error:
` Cannot find module '/Users/myuser/test-app/node_modules/@nexrender/action-encode/index.js' at webpackEmptyContext (webpack:///./node_modules/requireg/lib_sync?:2:10) at requireg`

**Comment by FinnFrotscher** [comment]
cant help you. i have mostly forked the repo. one of the reasons was issues with `requireg`

**Comment by morneluckybeard** [comment]
Would have loved to use the programmatic version but I ended up running the server (programmatically) and worker(using exec child_process) on the same machine. API seems to work fine with that setup. 

I think my scenario had something to do with the way electron bundled the required node_modules but I'm uncertain.

Thanks anyways!

**Comment by matt-cook** [comment]
Running into this while incorporating into a NextJS project as well.
